docs/design/proposals/encryption-with-vault-tokens.md

```diff
@@ -0,0 +1,125 @@
+
+## Dependencies
+
+- the name of the Kubernetes Secret needs to be known (configured in the
+  StorageClass, or somewhere else)
+- each tenant (Kubernetes namespace) should use the same name for the Secret
+  that contains the Vault token
+- the Ceph-CSI sidecar needs to store the namespace (and name of the Secret?)
+  together with the metadata of the PVC
+
+## Implementation Outline
+
+- when creating the PVC the Ceph-CSI provisioner needs to store the namespace
+  (and Kubernetes Secret?) of the PVC in its metadata
+  - stores the `csi.volume.owner` (name of Tenant) in the metadata of the
+    volume and sets it as `rbdVolume.Owner`
+- the Ceph-CSI node-plugin needs to request the token in the NodeStage CSI
+  operation and create/get the key for the PVC
+- the Ceph-CSI provisioner needs to request the token to delete the key for the
+  PVC in the VolumeDelete CSI operation
+
+### Fetching Tenant Tokens from Kubernetes
+
+CSI is an abstraction and should not communicate with Kubernetes or other
+Container Orchestration systems directly. Therefore, it is needed to
+communicate with a service that can provide the token from the tenants. This
+service can be provided by a sidecar, exposing an endpoint as a
+UNIX-Domain-Socket to communicate through.
+
+This sidecar does not exist at the moment. There are other features within
+Ceph-CSI that will benefit from this sidecar, e.g. Topology support. Because
+Ceph-CSI already uses a Kubernetes client to fetch details from Kubernetes, it
+should be acceptable to fetch the Vault Token configuration for the tenants the
+same way.
+
+## Configuration Details
+
+- the current KMS configuration file needs extensions for Vault Token support
+- introduce a new KMS-type: VaultTokenKMS (the current VaultKMS uses a
+  Kubernetes ServiceAccount)
+- configuration of the VaultTokenKMS can be very similar to VaultKMS for common
+  settings
+- the configuration can override the defaults for each tenant separately
+  - Vault Service connection details (address, TLS options, ...)
+  - default token (in case of a single token for every tenant)
+  - name of the Kubernetes Secret that can be looked up per tenant
+- tokens for each tenant can be added to the configuration (insecure & needs
+  admin involvement)
+- the configuration points to a Kubernetes Secret per tenant that contains the
+  token, but also can contain alternative options for the VaultTokenKMS service
+
+### Example of the KMS configuration file for Vault Tokens
+
+The configuration is available in the Ceph-CSI containers as
+`/etc/ceph-csi-encryption-kms-config/config.json`:
+
+```json
+{
+    "vault-with-tokens": {
+        "encryptionKMSType": "vaulttoken",
+        "vaultAddress": "http://vault.default.svc.cluster.local:8200",
+        "vaultCAFromSecret": "vault-ca",
+        "vaultCAVerify": "false",
+        "vaultToken": "@@B64_ENCODED_DEFAULT_TOKEN@@",
```

Comment:
I would suggest we move out all secrets from the config map to a secret within the same namespace as ceph-csi. This way, we map in the secret as well into (say) `/etc/ceph-csi-tokens.json` in the nodeplugin/provisioner deployments as an additional volume. It would keep it clean, as in secrets are in a kubernetes secret and not in a config map.

Reply:
Yeah, I'm not very fond if listing secrets in a ConfigMap. Ideally the tokens are not available as a file at all, and rotating tokens would just be automatic by issuing a new API call to get it.

The listing of tokens in the ConfigMap was the initial design, and was kept here for now. It would be the 1st thing to drop from the configuration file :smiley:

Reply:
Tokens are not stored in the ConfigMap anymore, #1759 does not read tokens from the ConfigMap (or file) either.